### DISCUSSION TITLE
handle.useQuery() should only be called in the setup() or a lifecycle hook.

### DISCUSSION THREAD

**Discussion by wucdbm** [urql-graphql/urql#2781] [Q&A]
As a continuation of https://github.com/FormidableLabs/urql/issues/2779

@kitten 

> Yes, but the main useClientHandle function has to be called in the setup method.

Yes, but the warning appears when you call any of the functions returned in the handle, not when you call useClientHandle itself. And since useClient is called on the first line, it will definitely warn about using use* functions properly in your setup. Hence, the if statement is redundant. More on what happens next below.

I am only unsure about `useEffect()` calls in `callUseQuery`, but having used it (in one place so far!) via getting a handle first, then calling `handle.useQuery` in a callback on click, it hasn't caused any issues.

But then that begs the question, why does `handle.useQuery` and `handle.useSubscription` exist in the first place, given that you must call it in the setup function context (**not in a callback!**), when you could simply call `useQuery` and `useSubscription` directly then (**Yes, I have read the documentation about it**, comment on that in the next paragraphs).

And if `useEffect` is to cause any issues when called in a callback past `setup()` context, then it makes more sense to get rid of `useQuery` and `useSubscription` in the handle as it's misleading as to what its purpose in life is.

But then again, **clearly the documentation states that this kind of usage (past setup() context) is advised** - https://formidable.com/open-source/urql/docs/basics/vue/#chaining-calls-in-vue-suspense

It does not matter if you have awaited a result, or called `handle.use*` in a callback, **you're already past setup() context**. This is also the reason why vue's ESLint rules warn you not to call any lifecycle hooks such as `watch` or `watchEffect` after an await statement in an async setup function - here are two examples

```
    async setup() {
        await Promise.resolve()

        // ESLint: `watch` is forbidden after an `await` expression.(vue/no-watch-after-await)
        watch(someRef, () => {

        })

        // ESLint: `watch` is forbidden after an `await` expression.(vue/no-watch-after-await)
        watchEffect()

        // Guess what'll happen if you call handle.useQuery where useQuery will call watchEffect -- no eslint warning, but the same reason for not doing so stands - rendering the below code taken from docs faulty:

        await Promise.resolve(); // NOTE: This could be any await call
        const result = await handle.useQuery({
```

The fact that it works and is documented like this only proves my point that it's OK to get a reference to a handle in setup() then call its members in a callback. **HOWEVER**

This basically renders what is being said in the documentation link provided above wrong. My usage as well. Calling `const result = await handle.useQuery({...` will internally call useEffect(), which by Vue's linters is forbidden after an await, since we're by then already past the lifecycle hooks context. See EDIT at the end for more context.

I just ran a basic example with reactive `variables`, executing a query in a callback via `handle.useQuery` and setting every new result's data to a ref I've setup earlier in the setup function and not only does it work just OK, it also reacts when I change the underlying query parameter via the reactive variables without calling the callback to fetch a new result and set a new data ref in my outside ref (ref in ref in order to test reactivity if computed vars change).

So IMO there's two sides here

- On one hand, this works just fine and watchEffect does not seem to cause any issues being called after setup()
- On the other hand, based on your comment about the canonical space to setup lifecycle hooks, even the advised usage in the documentation is wrong and the example and usage pattern have to be removed altogether as they do not adhere to what you just mentioned about lifecycle hooks - proof are vue's ESLint rules in the docblock above.

With all that being said, and having read the answer in the thread you linked, I am completely fine calling `client.query(...).toPromise` by hand and "typing" the query and variables for type-safety myself (context: we're using graphql-codegen and all but one of our queries are called via useWhateverQuery that internally calls useQuery with pre-set generic params for data type, variables type). **But** I also believe that if you must be strict as per your own comment, then `handle.useQuery` and `handle.useSubscription` **have to go** as they make no sense in this context and there's no other world where you would need to use them.

**EDIT** You can test that after an `await Xyz` you're outside of the scope of a setup function lifecycle by awaiting something and then calling `inject()` - it'll error hard on you:

```
async setup() {
        await Promise.resolve()

        inject('some key does not matter anymore as vue throws an error')
```

`Error: use* functions may only be called during the `setup()` or other lifecycle hooks.`

**Comment by wucdbm** [comment]
Edited `handle.useMutation` to `handle.useSubscription` at the end, was a brainfart

**Comment by kitten** [comment]
> I am only unsure about useEffect() calls in callUseQuery, but having used it (in one place so far!) via getting a handle first, then calling handle.useQuery in a callback on click, it hasn't caused any issues.

It can cause issues however because that usage explicitly triggers a query then only marks it as inactive on an unmount.

Basically these function exists so you can chain await calls of `await handle.useQuery`, however, for programmatic calls you'll either want to simply update a ref that provides variables and manipulate a ref that provides `pause`, or alternatively use `client.query(...).toPromise()` if you're only interested in a single result in no updates.

The constraint here exists because it sets up the effect chain, but even if that chain is set up, if it's set up in a callback then there's nothing but the unmount that will get rid of it.

Say, you have a form and a button. The button when it clicked calls `handle.useQuery`. The component will now accumulate queries infinitely and create new result refs and keep all of these queries active until it's unmounted, which is essentially a leak.
That's why we provide `pause` and why these effect chains are typically set up only in `setup`. To keep them to a minimum.

> And if useEffect is to cause any issues when called in a callback past setup() context

It doesn't necessarily cause an issue, and the warning is safe to ignore, if you only call `useQuery` _once_ outside of `setup` and then use the ref as intended. It's basically a guard rail to say that it's safe to call the method once before the component is actually mounted.

> It does not matter if you have awaited a result, or called handle.use* in a callback, you're already past setup() context.

Yes, this rule, afaik, is valid on its own because the `watch` call will not have access automatically to a teardown on unmount, which is why the `handle` will keep track of those: https://github.com/FormidableLabs/urql/blob/8fc678e8bf81c72d57f5cf27f03bf3b132dcf951/packages/vue-urql/src/useClientHandle.ts#L44

It basically is a rule that says that we shouldn't set up effect chains after having awaited, however, if multiple awaits are necessary, that's why the handle exists, because while queries in GraphQL can and should be merged into a single one, sometimes a component of course needs access to having set up a query, and one or more mutations and subscriptions

> it also reacts when I change the underlying query parameter via the reactive variables without calling the callback to fetch a new result and set a new data ref in my outside ref

so basically, it works, and the chains will all correctly work and stop on unmount, but they'll also never stop. That means, when you set up more on a user interaction, the prior chains do not stop and you'll create a leak. Meaning, all the queries and subscriptions you'll set up will remain active.

Hope that clarifies it

**Comment by wucdbm** [comment]
Right, I can agree on the leak part. Let me examine your reply more thoroughly these days to better understand what's going on with the stops as well.

Other than that, in this context the error message starts to smell - would you consider adding this context to it - surely I'm not the first person wondering why this use case is considered illegal. Possibly also a warning in the documentation right after the handle example, describing exactly this reason as to why that shouldn't be done for a one-off queries?

**Reply by kitten** [reply]
That makes sense 👍 I'll see how we can improve that message
